lib/attribute.js

```diff
@@ -0,0 +1,356 @@
+// Highly modeled after the interface for YUI's Attribute.
+// Though much simpler internally. 
+
+var ObjectUtils = require('./utilities/objects');
+var Lang = require('./utilities/lang');
+var EventTarget = require('./eventTarget');
+var augment = require('./utilities/augment');
+
+var OPTS = {
+  GETTER: 'getter',
+  SETTER: 'setter',
+  READ_ONLY: 'readOnly',
+  INIT_ONLY: 'initOnly',
+  WRITE_ONCE: 'writeOnce',
+  VALIDATOR: 'validator',
+  VALUE: 'value',
+  VALUE_FN: 'valueFn'
+};
+
+/* 
+private properties related to an attr are written to _attrState:
+_setValue - takes precedence over value and valueFn, which shant be changed.
+_valueWasSet - whether _setValue has been set
+_writtenOnce - whether it's been set at least once
+ */
+
+var VALID_OPTIONS = ObjectUtils.values(OPTS);
+
+var validateConfig = function(cfg) {
+  for (var key in cfg) {
+    if (cfg.hasOwnProperty(key) &&
+      VALID_OPTIONS.indexOf(key) === -1) {
+      console.error('Invalid config parameter: ' + key);
+      return false;
+    }
+  }
+
+  if (cfg[OPTS.READ_ONLY] && !(OPTS.VALUE in cfg)) {
+    console.error('readOnly attribute ' + name + ' must have an initial value.');
+    return false;
+  }
+
+  return true;
+};
+
+var stringToInstanceMethod = function(functionName) {
```

Comment:
Good question.

I'm personally of the opinion that we should be documenting all of our code as we add it. This isn't a public facing section of the codebase, but having at least some basic commentary explaining at a high level what everything is doing would be nice. Indicating expected parameter types and return types is also something that I think is important since it saves me from having to go in and infer that from the code.

... that being said, I'm also well aware that we have a deadline that is rapidly approaching, and I don't want to be the road block to getting this code merged in. We can table the documentation issue for now - expect lots of questions from anyone who has to go through this though :).